### ISSUE TITLE
Too many requests accessing the tasklist API

### ISSUE THREAD

**Issue by zambrovski** [camunda-community-hub/camunda-tasklist-client-java#11]
Hi folks,

I'm trying to use the tasklist-client to connect to the SaaS account and getting the following exception:

```
Caused by: io.camunda.tasklist.exception.TaskListException: Error 429 obtaining access token : Too Many Requests
	at io.camunda.tasklist.auth.SaasAuthentication.authenticate(SaasAuthentication.java:88) ~[camunda-tasklist-client-java-1.5.2.jar:na]
	at io.camunda.tasklist.CamundaTaskListClient$Builder.build(CamundaTaskListClient.java:212) ~[camunda-tasklist-client-java-1.5.2.jar:na]
	at io.holunda.camunda.worker.impl.Camunda8Configuration.camundaTaskListClient(Camunda8Configuration.kt:62) ~[classes/:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.3.23.jar:5.3.23]
	... 48 common frames omitted

```

I verified the clientId, clientSecret, and Url, they seem to be ok.

Any ideas?

Camunda Tasklist Client: 1.5.2
Zeebe Cluster Version 8.1.4
OS: Ubuntu 22.04
Java: 17

Cheers,

Simon

**Comment by chDame** [comment]
Hi Simon,

I just created a new cluster 8.1.5 and tried this client version without problem. The error implies that too many request are sent simultaneously and the server is protecting against an eventual ddos attack.

Is it possible that your clientId and clientSecret have been exposed ? May your code send too many request simultaneously?

**Comment by zambrovski** [comment]
This is what I don't understand. The only think I did was to create the following in Spring Boot configuration:

```

class Camunda8Configuration {

  @Value("\${tasklist.url}")
  private lateinit var taskListUrl: String

  @Value("\${operate.url}")
  private lateinit var operateUrl: String

  @Bean
  @ConditionalOnMissingBean
  fun camunda8processStarter(zeebeClient: ZeebeClient): ProcessStarter = Camunda8ZeebeProcessStarter(zeebeClient)

  @Bean
  fun tasklistSaasAuthentication(zeebeClientCloudConfigurationProperties: ZeebeClientConfigurationProperties) = TasklistSaasAuthentication(
    zeebeClientCloudConfigurationProperties.cloud.clientId,
    zeebeClientCloudConfigurationProperties.cloud.clientSecret
  )

  @Bean
  fun operateSaasAuthentication(zeebeClientCloudConfigurationProperties: ZeebeClientConfigurationProperties) = OperateSaasAuthentication(
    zeebeClientCloudConfigurationProperties.cloud.clientId,
    zeebeClientCloudConfigurationProperties.cloud.clientSecret
  )


  @Bean
  fun camundaTaskListClient(auth: TasklistSaasAuthentication): CamundaTaskListClient = CamundaTaskListClient
    .Builder()
    .taskListUrl(taskListUrl)
    .authentication(auth)
    .build()


  @Bean
  fun operateClient(auth: OperateSaasAuthentication): CamundaOperateClient = CamundaOperateClient
    .Builder()
    .operateUrl(operateUrl)
    .authentication(auth)
    .build()

  @Bean
  fun taskAccessor(
    camundaTaskListClient: CamundaTaskListClient,
    operateBetaClient: CamundaOperateClient
  ) = TaskAccessor(
    taskListClient = camundaTaskListClient,
    operateBetaClient = operateBetaClient
  )
}

```

**Comment by zambrovski** [comment]
Christophe,

there is nothing broken in this code. There is definitely a problem of rate limiting on the SaaS side. Since there is no documentation on that this issue was an attempt to find out how 
often I'm allowed to connect and how the rate limiting is working.

Without changing anything this code worked today. Probably it will stop if I hit the rate limit again.

About your question: the code is executed exactly once per application start. Yesterday I had to restart the application maybe 10-15 times and that led me to the rate limit. I also debugged through the entire code of the client - it seems to be the SaaS restriction.


P.S. probably I'm not the only one who experienced this error.


Cheers,

Simon
